### PR TITLE
Threads

### PR DESCRIPTION
**Pull request by DonDuala**
Reintroduces cancel logic to only attempt to cancel if there is an order of that asset open. Kills gamma threads after max of rerun time. Limits # of open gamma threads to the number of fills during a single scalperWindow

### CODE REVIEW

File: src/scalper.ts [hunk 1]
```diff
@@ -462,7 +463,11 @@ export class Scalper {
     } catch (err) {
         console.log(this.symbol, "Gamma Ask Error", err);
         console.log(this.symbol, "Gamma Ask Error Details",err.stack);
-      }
+    }
+    
+    // Sleep for the max time of the reruns then kill thread
+    await sleepExact((1 + percentDrift) * scalperWindow);
+    return;
```
Comment (brittcyr) [inline src/scalper.ts:470]: What is this supposed to do? The code should just exit normally already

Comment (DonDuala) [reply src/scalper.ts:470]: True will update to removeEventListeners that is the piece that lives on within a thread if we don't receive a fill

File: src/scalper.ts [hunk 2]
```diff
@@ -467,7 +467,8 @@ export class Scalper {
     
     // Sleep for the max time of the reruns then kill thread
     await sleepExact((1 + percentDrift) * scalperWindow);
-    return;
+    console.log(this.symbol, "Remove stale gamma fill listener", gammaBidID, gammaAskID)
+    fillFeed.removeEventListener('message', gammaFillListener);
```
Comment (DonDuala) [inline src/scalper.ts:471]: Still testing this to make sure it doesn't intercept and kill a gammaFillListener from a current thread. Not sure the scope here if you know

File: src/main.ts [hunk 3]
```diff
@@ -43,29 +44,52 @@ async function main() {
...
+    setInterval(async () => {
+      console.log('Rerun BTC Risk Manager', new Date().toUTCString());
+      try {
+        await btcRouter.refresh_dips();
+        btcRouter.run_risk_manager();
+      } catch (err) {
+        console.log(err);
+        console.log(err.stack);
+      }
+    }, 1_000 * (scalperWindow + (((Math.random()*2) - 1) * scalperWindow * percentDrift))
```
Comment (brittcyr) [inline src/main.ts:78]: You still want them all on the same interval so they dont drift, just want them smeared. Should be able to do it with sleeps right after run_risk_manager for each

Comment (DonDuala) [reply src/main.ts:78]: Will this trim down the last products total run time though? ie. Will ETH gamma scalper only be in reality running for 9 mins

Comment (brittcyr) [reply src/main.ts:78]: setInterval only sets when to start, not how long to run for. It allows overlapping calls

File: src/scalper.ts (hunk 2 above)
Comment (DonDuala) [reply src/scalper.ts:471]: image
image
correctly removed the old ones
image

**Review by brittcyr — APPROVED** [review]

**Pull request merged**